app/template/page.tsx

```diff
@@ -0,0 +1,28 @@
+import { Metadata } from 'next';
+import PageClient, { IServerDataSchema } from './page.client';
+
+export const metadata: Metadata = {
+  title: 'EDPN Template',
+  description: 'Elite Dangerous Pilots Network',
+  icons: 'EDPN_logo_dark_background.png',
+};
+
+async function getData() {
```

Comment:
With this API call being used to get commodity data for a dropdown on about 4-5 pages and will always be the same, does it make sense to move it all into a hook or a faux hook? I'm just thinking that this function will be copied pretty much verbatim for each page, so maybe we could simplify it further by having a useGetCommodities() we can just drop onto a page (or a non hook function if we think it might get used outside of a tsx component)

I appreciate it's only an example, but you highlighted the point quite well in this example where you've made 2 identical queries client and server side. Which I get this is niche, it could be nice to be able to just call `getCommodities(paramsObject)` and have a reusable hook/function/whatever we can call so we don't need to start hunting around for urls, cache times other developers have used, etc

Or maybe go more generic with a getApiRequest function where you specify the endpoint and the params?

I've probably not spent enough time thinking about it, my line of thinking is simply that every time I created a page with a commodity drop down, I'd need to search the codebase for the URL, the cache time, and probably copypaste a bunch of code, or if someone suggests increasing cache time we would need to change it in several place - which just shouts to me that it should be abstracted. 

Perhaps it just needs a bit of further discussion, or maybe we just accept that it's not enough of a hassle to warrant extra work and we can just live with it? Unsure tbh. Thoughts?

Reply:
@chrismou I have a working change locally that will implement swr based hooks (https://github.com/ed-pilots-network/frontend/issues/105), with all the open PRs, I have been holding on to that change and waiting for this one to be merged as I saw @aslink87 has moved the API call out and will make what I am doing even simpler.

Hook Example - app/_hooks/_api
```javascript
import useSWR from 'swr';
import { ICommodity } from '@/app/_types';

export default function useCommodity(name: string) {
  const { data, error, isLoading } = useSWR(
    `${process.env.NEXT_PUBLIC_MOCK_API_URL}/api/v1/trade/commodity/${name}`,
  );

  return {
    commodity: data as ICommodity,
    isLoading,
    isError: error as Error,
  };
}
```

Using the hook

```javascript
  import useCommodity from '@/hooks/api/useCommodity';
  const { commodity, isLoading, isError } = useCommodity('Tritium');

  const renderCommodityData = () => {
    if (isLoading) {
      return <Spinner />;
    }
    if (isError) {
      return (
        <Alert status="error">
          <AlertIcon />
          Failed to fetch commodity data from staging server on client side
        </Alert>
      );
    }
    if (commodity) {
      return <div>Commodity Name: {commodity.commodityName}</div>;
    }
```

Reply:
Nextjs recommends swr for client-side fetching [here](https://nextjs.org/docs/pages/building-your-application/data-fetching/client-side#client-side-data-fetching-with-swr).

If that's the case, then we would continue fetching the server data that's needed in the pages within the page.tsx similar to how I've done it here. We could use swr for the client side form submissions.

Reply:
I like where @chrismou was heading in his comment. Correct me if i'm wrong but all our server fetches are for arrays of data used in form elements. I don't see that changing. So we could look into extracting that server call and calling it in each page.tsx where it is needed. `getFormProps(url)`.

Then if we go with swr, we could have a universal submitForm function to return the form's result.

Reply:
@aslink87 Extracting that server call to where? like as a swr hook? I don't think we can use hooks in server pages, but if you mean we should move it to a common place so other pages can use the same method (not a swr hook), then yeah I am aligned. Client side fetches will be made re-usable with swr based hooks and server side can be just functions packed under some server side api utility. Similar to what you have in this change and what I have in the playground API  - https://github.com/ed-pilots-network/frontend/blob/development/app/playground/_api/index.tsx

Reply:
Tomorrow I'll move the server call to a more universal location. @nitishdhar Do you want to add your SWR implementation right into this branch? would be nice to have a complete picture here before merging into dev.

Reply:
Will do 👍

Reply:
extracted api calls to universal functions: f00017268f18037582c76acd74427b16dae83bb9

Reply:
@aslink87 I have implemented swr in your branch. Couple of things - I noticed you have made the fetch as general calls so I kept the new hook also under the same library file. An alternate would be to move something like `client-api-calls` under `hooks` folder as all client fetches will now be hooks and keep server side calls under `_lib/server-api-calls`. I did it but then realized there's only one for both for now and that might not change unless out fetch patterns change so reverted to just that one library file. Also, replaced the custom response with a specific type with updated comments to replace with relevant resource type - https://github.com/ed-pilots-network/frontend/pull/119/commits/2cd2090e4ce6e5b6f80ee77d0eb43e68fc54d876